### PR TITLE
Update CupertinoSlidingSegmentedControl

### LINKED ISSUE #62063: CupertinoSlidingSegmentedControl: Calculated width doesn't account for highlighted style
`CupertinoSlidingSegmentedControl` currently determines its width based on the width of the widest child widget. This calculation doesn't account for the (slight) increase of the child's width when it's selected:

Screen Recording 2020-07-22 at 12 29 33 PM mov

This small change in width happens regardless of which child is the widest:

Screen Recording 2020-07-22 at 1 10 14 PM mov

---

Additionally, I think the overall width semantics don't match the iOS spec. `CupertinoSlidingSegmentedControl` currently sizes itself to be as small as it needs to be in order to fit all of the children (set to be the same width as the biggest child).

It seems like the correct semantics would be that its width is the maximum allowed by its parent. From the docs for [`UISegmentedControl`](https://developer.apple.com/documentation/uikit/uisegmentedcontrol):

```
The UISegmentedControl object automatically resizes segments to fit proportionally within their 
superview unless they have a specific width set.
```

### LINKED ISSUE #63338: CupertinoSlidingSegmentedControl crash : AnimationController.animateTo() called after AnimationController.dispose()
App is crashing when using CupertinoSlidingSegmentedControl  (on Iphone 6s). I'm doing some heavy work (loading maps with many polygons) while using this slider, and when I swipe it back really soon, it crashes (not every time)

stack trace:
```
The following assertion was thrown while handling a gesture:
AnimationController.animateTo() called after AnimationController.dispose()
AnimationController methods should not be used after calling dispose.
'package:flutter/src/animation/animation_controller.dart':
Failed assertion: line 512 pos 7: '_ticker != null'

When the exception was thrown, this was the stack: 
#2      AnimationController.animateTo (package:flutter/src/animation/animation_controller.dart:512:7)
#3      _SegmentedControlState.pressed= (package:flutter/src/cupertino/sliding_segmented_control.dart:384:36)
#4      _RenderSegmentedControl.pressedIndex= (package:flutter/src/cupertino/sliding_segmented_control.dart:662:11)
#5      _RenderSegmentedControl._onCancel (package:flutter/src/cupertino/sliding_segmented_control.dart:746:5)
#6      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:182:24)
...
Handler: "onCancel"
Recognizer: HorizontalDragGestureRecognizer#c3c53
  start behavior: start
```

flutter doctor:
> [✓] Flutter (Channel stable, v1.17.5, on Mac OS X 10.15.3 19D76, locale en-GB)
>  
> [✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
> [✓] Xcode - develop for iOS and macOS (Xcode 11.6)
> [✓] Android Studio (version 3.6)

### LINKED ISSUE #73362: _RenderSegmentedControl's computeDryLayout function uses wrong constraints
Captura de pantalla 2021-01-05 a las 22 47 46

Related PR https://github.com/flutter/flutter/pull/70656

CODE WHERE THE PROBLEM IS
https://github.com/flutter/flutter/blob/542084d08830a6d7d64c70e4a367eaed4ca3c74a/packages/flutter/lib/src/cupertino/sliding_segmented_control.dart#L850-L858

Inside `computeDryLayout` method, `_computeOverallSizeFromChildSize` is called. This method uses the constraints from the RendexBox and it will through an error as computeDryLayout is expected to be called before the renderbox has been laid out. In computeDryLayout it should use the constraints passed as param. 

I think the fix is an easy change 
```dart
  Size _computeOverallSizeFromChildSize(Size childSize, BoxConstraints constraints) {
    return constraints.constrain(Size(childSize.width * childCount + totalSeparatorWidth, childSize.height));
  }

  @override
  Size computeDryLayout(BoxConstraints constraints) {
    final Size childSize = _calculateChildSize(constraints);
    return _computeOverallSizeFromChildSize(childSize, constraints);
  }

```

VERSION:  Flutter (Channel dev, 1.26.0-1.0.pre, on macOS 11.1 20C69 darwin-x64, locale es-ES)

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Keep an offstage copy for each segment, allowing the parent widget to get an overall size of all children that is `groupValue`-invariant.

Add missing animation: the content of the selected segment shrinks (up to 95% of the original size) when the user presses on it and hold.

Make the render object more lightweight.

- Fixes https://github.com/flutter/flutter/issues/62063
- Fixes https://github.com/flutter/flutter/issues/73362
- Fixes https://github.com/flutter/flutter/issues/63338

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/sliding_segmented_control.dart [hunk 1]
```diff
@@ -94,32 +268,26 @@ class _FontWeightTween extends Tween<FontWeight> {
...
 /// the keys will determine the order of the widgets in the segmented control.
 ///
-/// When the state of the segmented control changes, the widget calls the
-/// [onValueChanged] callback. The map key associated with the newly selected
-/// widget is returned in the [onValueChanged] callback. Typically, widgets
-/// that use a segmented control will listen for the [onValueChanged] callback
-/// and rebuild the segmented control with a new [groupValue] to update which
-/// option is currently selected.
+/// The widget calls the [onValueChanged] callback *when a valid user gesture
+/// completes on a unselected segment*. The map key associated with the newly
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:272]: a => an

File: packages/flutter/lib/src/cupertino/sliding_segmented_control.dart [hunk 2]
```diff
@@ -357,111 +480,236 @@ class _SegmentedControlState<T> extends State<CupertinoSlidingSegmentedControl<T
...
+        index = numOfChildren - 1 - index;
+        break;
+    }
+
+    assert(index < widget.children.length);
+    for (final T key in widget.children.keys) {
+      if (index == 0)
+        return key;
+      index -= 1;
+    }
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:514]: Couldn't you do `return widget.children.keys.elementAt(index);` and remove the assert and return below this?  Or maybe I'm missing something...

File: packages/flutter/lib/src/cupertino/sliding_segmented_control.dart [hunk 3]
```diff
@@ -357,111 +480,236 @@ class _SegmentedControlState<T> extends State<CupertinoSlidingSegmentedControl<T
...
+  void onCancel() {
+    if (isThumbDragging) {
+      _playThumbScaleAnimation(isExpanding: true);
+    }
+
+    onPressedChangedByGesture(null);
+    _startedOnSelectedSegment = null;
+  }
+
+  // The segment the sliding thumb is currently located, or animating to. It may
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:625]: "located" => "located at"

File: packages/flutter/lib/src/cupertino/sliding_segmented_control.dart [hunk 4]
```diff
@@ -842,158 +945,113 @@ class _RenderSegmentedControl<T> extends RenderBox
...
+      final _SegmentedControlContainerBoxParentData childParentData = selectedChild.parentData! as _SegmentedControlContainerBoxParentData;
+      final Rect newThumbRect = _kThumbInsets.inflateRect(childParentData.offset & selectedChild.size);
+
+      // Update thumb animation's tween, in case the end rect changed (e.g., a
+      // new segment is added during the animation).
+      if (state.thumbController.isAnimating) {
+        final Animatable<Rect?>? thumbTween = state.thumbAnimatable;
+        if (thumbTween == null) {
+          // This is the first frame of the animation.
+          // TODO(LongCatIsLooong): `currentThumbRect` can be outside of the
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:1013]: Did you intend to fix this TODO before merging?

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Looks much more solid to me.  Just some nits.

**Comment by LongCatIsLooong** [comment]
Addressed the TODO and added a new TODO for the animation.

Replaced the global keys with `ValueKey`s because they're not used.

File: packages/flutter/lib/src/cupertino/sliding_segmented_control.dart [hunk 5]
```diff
@@ -357,111 +469,233 @@ class _SegmentedControlState<T> extends State<CupertinoSlidingSegmentedControl<T
...
       return;
-    _animateHighlightController(at: newValue, forward: true);
-    _animateHighlightController(at: _highlighted, forward: false);
-    _highlighted = newValue;
+    setState(() { highlighted = newValue; });
+    // Additionally, start the thumb animation if the highlighted segment
+    // changes. If the thumbController is already running, the render object's
+    // paint method will create a new tween to drive the animation with.
+    // TODO(LongCatIsLooong): the current thumb will be painted at the same
+    // location twice before and after the new animation starts.
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:537]: Nit: Maybe create an issue for this unless you're not worried about losing it.

**Review by justinmc — APPROVED** [review]
The new thumb-bounds fix LGTM 👍

**Pull request merged**